### PR TITLE
Plugins: Pass OAuth Token to CallResource Function

### LINKED ISSUE #45938: Grafana oauthPassThru external plugins?
Does Grafana `oauthPassThru` only work for datasources embedded within Grafana itself? Looking at the code this appears to be the case. Is there a specific reason for this? The documentation here https://grafana.com/docs/grafana/latest/developers/plugins/add-authentication-for-data-source-plugins/#forward-oauth-identity-for-the-logged-in-user doesn't make any mention of this limitation but the code example it provides isn't compatible with the grafana-plugin-sdk-go module (i.e. there isn't a "Headers" property on the "DataQuery" type so the code example given doesn't even compile).

### PR DESCRIPTION
**Pull request by bradenhs**
**What this PR does / why we need it**:

This PR ensures there is OAuthPassThru support for the `CallResource` function in plugins in addition to the `QueryData` function. Note: This PR doesn't tackle getting the appropriate headers forwarded for additional endpoints such as `CheckHealth`. I probably won't tackle that work personally as I don't need it but I expect folks will want that as well.

**Which issue(s) this PR fixes**:

Fixes https://github.com/grafana/grafana/issues/45938

Provided this general approach looks good I'd be happy to figure out how to add tests for this.

### CODE REVIEW

**Review by wbrowne — CHANGES_REQUESTED** [review]
Thanks for your contribution @bradenhs, but unfortunately this breaks the following endpoints:
- `/plugins/:pluginId/resources`
- `/api/plugins/:pluginId/resources/*`

The reason being that these have no strict dependency on a data source existing for a given plugin. This is quite confusing though as the code does pass around a `dsUID`, and it's not made clear that this can be empty. In the future, these should likely be two separate flows to avoid confusion.

I think for now though we could simply do:

```golang
...
clonedReq.URL = urlPath

if dsUID != "" {
	ds, err := hs.DataSourceCache.GetDatasourceByUID(c.Req.Context(), dsUID, c.SignedInUser, c.SkipCache)
	if err != nil && !errors.Is(err, models.ErrDataSourceNotFound) {
		handleCallResourceError(err, c)
		return
	}

	if hs.DataProxy.OAuthTokenService.IsOAuthPassThruEnabled(ds) {
		if token := hs.DataProxy.OAuthTokenService.GetCurrentOAuthToken(c.Req.Context(), c.SignedInUser); token != nil {
			clonedReq.Header.Add("Authorization", fmt.Sprintf("%s %s", token.Type(), token.AccessToken))

			idToken, ok := token.Extra("id_token").(string)
			if ok && idToken != "" {
				clonedReq.Header.Add("X-ID-Token", idToken)
			}
		}
	}
}

if err = hs.makePluginResourceRequest(c.Resp, clonedReq, pCtx); err != nil {
...
```

WDYT?

**Comment by bradenhs** [comment]
Thanks @wbrowne! I made your suggested change.

File: pkg/api/plugins.go [hunk 1]
```diff
@@ -506,6 +506,26 @@ func (hs *HTTPServer) callPluginResource(c *models.ReqContext, pluginID, dsUID s
...
 
+	if dsUID != "" {
+		ds, err := hs.DataSourceCache.GetDatasourceByUID(c.Req.Context(), dsUID, c.SignedInUser, c.SkipCache)
+
+		if err != nil && !errors.Is(err, models.ErrDataSourceNotFound) {
+			handleCallResourceError(err, c)
+			return
+		}
+
+		if ds != nil && hs.DataProxy.OAuthTokenService.IsOAuthPassThruEnabled(ds) {
```
Comment (wbrowne) [inline pkg/api/plugins.go:517]: No need to `nil` check ds here as we presume it's okay when checking for `err` above

File: pkg/api/plugins.go [hunk 2]
```diff
@@ -506,6 +506,26 @@ func (hs *HTTPServer) callPluginResource(c *models.ReqContext, pluginID, dsUID s
 	}
 	clonedReq.URL = urlPath
 
+	if dsUID != "" {
+		ds, err := hs.DataSourceCache.GetDatasourceByUID(c.Req.Context(), dsUID, c.SignedInUser, c.SkipCache)
+
+		if err != nil && !errors.Is(err, models.ErrDataSourceNotFound) {
```
Comment (wbrowne) [inline pkg/api/plugins.go:512]:
```suggestion
		if err != nil {
			if errors.Is(err, models.ErrDataSourceNotFound) {
				c.JsonApiErr(404, "Datasource not found", err)
				return
			}

			c.JsonApiErr(500, "Failed to get datasource", err)
			return
		}
```

**Review by wbrowne — CHANGES_REQUESTED** [review]
Great stuff. Just noticed a couple things!

**Comment by bradenhs** [comment]
Thanks for the feedback! I made your suggested updates.

**Review by wbrowne — APPROVED** [review]
LGTM. Thanks a lot @bradenhs!

**Pull request merged**